### PR TITLE
Use new health check to verify if a server is up

### PR DESCRIPTION
**Pull request by jmoguilevsky**
In this PR:
- Use the new health check to verify if a server is up, only connect to it if every server is healthy
- Separate the CatalystClientBuilder to a helper so that it's more testeable
- Change chai +mocha tests for jest

### CODE REVIEW

File: src/utils/CatalystClientBuilder.ts [hunk 1]
```diff
@@ -35,114 +39,12 @@ export async function clientConnectedToCatalystIn(
...
-  const shuffledPreKnownServers = shuffleArray(knownServers)
-
-  // Ask N of them for their list
-  const approvedServersList = await Promise.all(
-    shuffledPreKnownServers
-      .slice(0, requiredAmountOfLists + 3)
-      .map((server) => server.address)
-      .map((address) => catalystListFetch(address, origin))
-  )
+    const result = await client.fetchPeerHealth()
```
Comment (nchamo) [inline src/utils/CatalystClientBuilder.ts:42]: Consider adding a timeout here. It should have been there in the first place, but it's a good time to add it

File: src/utils/CatalystClientBuilder.ts [hunk 2]
```diff
@@ -13,17 +14,20 @@ export async function clientConnectedToCatalystIn(
...
+
   let list: string[]
   if (noContractList) {
     list = noContractList
   } else {
     console.warn('Falling back to the smart contract to get an updated list of active servers')
     list = await getApprovedListFromContract(network)
   }
+
   const shuffled = shuffleArray(list)
```
Comment (nchamo) [inline src/utils/CatalystClientBuilder.ts:26]: Weird thought, and I wanted to see your opinion. Do you think there is a good/non-expensive way to prioritize up to date servers and de-prioritize ones that have not been updated for a while?

Might make sense to tackle it in another issue though

**Review by nchamo — COMMENTED** [review]
Small review haha

File: src/utils/CatalystClientBuilder.ts (hunk 2 above)
Comment (jmoguilevsky) [reply src/utils/CatalystClientBuilder.ts:26]: The list is 12 peers long, so any logic will be non expensive, but I would do it in another PR

File: test/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    "module": "commonjs", /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'. */
+    "declaration": true, /* Generates corresponding '.d.ts' file. */
+    "declarationMap": true, /* Generates a sourcemap for each corresponding '.d.ts' file. */
+    "sourceMap": true, /* Generates corresponding '.map' file. */
+    "types": [
+      "node",
+      "jest"
+    ],
+    "esModuleInterop": true, /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */
+    "skipLibCheck": true, /* Skip type checking of declaration files. */
```
Comment (agusaldasoro) [inline test/tsconfig.json:13]: Why do we need this?

File: test/utils/CatalystClientBuilder.spec.ts [hunk 4]
```diff
@@ -1,106 +1,113 @@
...
-          fetchCount++
-          return Promise.resolve([])
-        }
-      })
-
-      expect(fetchCount).to.equal(0)
-      expect(result).to.be.undefined
+import { clientConnectedToCatalystIn } from '../../src/utils/CatalystClientBuilder'
+import * as catalystList from '../../src/utils/catalystList'
+// import { CatalystClient } from '../../src/CatalystClient'
```
Comment (agusaldasoro) [inline test/utils/CatalystClientBuilder.spec.ts:3]: Forgot to remove this line

File: test/utils/CatalystClientBuilder.spec.ts [hunk 5]
```diff
@@ -1,106 +1,113 @@
...
+
+const server1 = 'server1'
+const healthyServer = 'server2'
+const server3 = 'server3'
+
+const mockedServerList = [server1, healthyServer, server3]
+jest.mock('../../src/CatalystClient', () => ({
+  CatalystClient: jest
+    .fn()
+    .mockImplementationOnce(() => {
```
Comment (agusaldasoro) [inline test/utils/CatalystClientBuilder.spec.ts:16]: I'm not sure how this mock is being used in the tests

File: test/utils/catalystList.spec.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
+import { getUpdatedApprovedListWithoutQueryingContract } from '../../src/utils/catalystList'
+
+describe('Catalyst Client Builder', () => {
+  describe('getUpdatedApprovedListWithoutQueryingContract', () => {
+    const ORIGIN = 'origin'
+    const REQUIRED_LISTS = 2
+    const [ADDRESS1, ADDRESS2, ADDRESS3] = ['http://test1.com', 'http://test2.com', 'http://test3.com']
+
+    it('When the amount of known servers on the list is less than required, then no fetching is performed and nothing is returned', async () => {
```
Comment (agusaldasoro) [inline test/utils/catalystList.spec.ts:9]:
```suggestion
    it('When the amount of known servers on the list is more than required, then no fetching is performed and nothing is returned', async () => {
```

File: package-lock.json [hunk 7]
```diff
@@ -1,257 +1,10219 @@
 {
   "name": "dcl-catalyst-client",
   "version": "0.0.0-development",
-  "lockfileVersion": 1,
+  "lockfileVersion": 2,
```
[the remark this answers is not in the document]
Comment (jmoguilevsky) [reply package-lock.json:4]: Good catch, I don't think so, will revert to NPM 6

File: test/utils/catalystList.spec.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        list: [{ address: ADDRESS1 }, { address: ADDRESS2 }, { address: ADDRESS3 }],
+        fetchApprovedCatalysts: () => {
+          if (fetchCount++ === 0) {
+            return Promise.resolve(undefined)
+          }
+          return Promise.resolve([ADDRESS1])
+        }
+      })
+
+      expect(fetchCount).toEqual(3)
```
Comment (agusaldasoro) [inline test/utils/catalystList.spec.ts:37]: If only the first one failed, and the second one retrieved an address then why got the third one called too?

File: test/utils/catalystList.spec.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+          fetchCount++
+          return Promise.resolve(undefined)
+        }
+      })
+
+      expect(fetchCount).toEqual(3)
+      expect(result).toBeUndefined()
+    })
+
+    it(`When the intersection is empty, then nothing is returned`, async () => {
```
Comment (agusaldasoro) [inline test/utils/catalystList.spec.ts:55]: Let's clarify about which intersection we are talking about

File: package-lock.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (jmoguilevsky) [reply package-lock.json:4]: Just checked, it's backwards compatible, and it's the same version of node, so yes, we're ready
image

File: package.json [hunk 10]
```diff
@@ -45,24 +45,23 @@
...
-    "dcl-catalyst-commons": "^4.0.0",
+    "dcl-catalyst-commons": "^4.0.1",
     "dcl-crypto": "^2.2.0",
     "form-data": "^4.0.0"
   },
   "devDependencies": {
     "@types/chai": "^4.2.11",
     "@types/chai-as-promised": "^7.1.2",
+    "@types/jest": "^26.0.23",
+    "@types/mocha": "^8.2.0",
```
[the remark this answers is not in the document]
Comment (jmoguilevsky) [reply package.json]: No, removed the dependencies for mocha and chai (and their typings)

**Review by agusaldasoro — APPROVED** [review]
Great work!

**Pull request merged**